polkadot/runtime/westend/src/lib.rs

```diff
@@ -2064,7 +2068,7 @@ sp_api::impl_runtime_apis! {
 			parachains_runtime_api_impl::minimum_backing_votes::<Runtime>()
 		}
 
-		fn para_backing_state(para_id: ParaId) -> Option<polkadot_primitives::async_backing::BackingState> {
+		fn para_backing_state(para_id: ParaId) -> Option<polkadot_primitives::vstaging::async_backing::BackingState> {
```

Comment:
We have everything in vstaging, but don't bump the version number of the API.

We should not change the APIs without bumping the version. What should be possible (if I remember correctly) is that we implement v10 still returning the old types and on v11 we change the interface of the function to the new types. 

What should then be possible is that chains can choose to still implement the old interface, by staying at version 10. 

@tdimitrov is what I am saying actually true?

Reply:
Yes, that's correct. If the signature of the method is being changed `changed_in` is used to rename the method or even better -  just introduce another method to avoid confusion.
`changed_in` is a bit confusing as it will create `para_backing_state_BEFORE_V12` (the old method) and `para_backing_state` (the new method). It's a bit ugly.

But since the type is the same, why move it to vstaging on the first place? Can't it just stay in `v7` and not touch the runtime api?

Reply:
In my ideal world, we would introduce new functions and deprecate the old ones returning the old format. Removing them eventually, once we are pretty sure nobody is using them anymore. 

@tdimitrov If I understand correctly, if we use `changed_in`, we are actually still breaking backwards compatibility, because the name of the function changes? Or is that somehow hidden?

Reply:
I think we just need the old candidate descriptor and receipt structures for collators to be able to create them until the feature bit for v2 is turned on in production. 

In this PR I duplicated structures and put all of that in staging until node side is also moved to new primitives.

In an ideal world we should have different runtime API version with the new data structures, but in this case they are binary compatible. 

We should release all of these primitive changes as v8 and dont bump the runtime API version.

Reply:
> I think, in the end state we will only keep the old candidate descriptor and receipt structures in v7 for collators to be able to create them until the feature bit for v2 is turned on in production.

Actually do we need to follow some deprecation process? I would assume this is very low level stuff that is not supposed to be used by parachain builders directly. So just removing them plus major bump on crates.io would be just fine

Reply:
>  if we use changed_in, we are actually still breaking backwards compatibility, because the name of the function changes? Or is that somehow hidden?

Effectively the old method gets renamed ([here](https://github.com/paritytech/polkadot-sdk/blob/d1b0ef76a8b060437ec7dfc1bf6b400626cd6208/substrate/primitives/api/proc-macro/src/decl_runtime_apis.rs#L409)) and this needs to be handled on the node side like it's done here:
https://github.com/paritytech/polkadot-sdk/blob/bfb241d7f3e29152b9fe0ab3a0364ee0c89ec65c/cumulus/client/collator/src/service.rs#L199

It's a bit messy but the benefit is that we can use the new name for the new method and let the old one disappear. The downside is the additional handling I mentioned.

Reply:
I'd much rather not have duplicated runtime APIs if the inputs and outputs are backwards compatible in scale encoding (and especially since the implementations are the same).

Reply:
We are changing types of a public interface ... even if binary compatible, this should really cause a version bump. Changing the types in a public interface is already nasty, doing that even without a version bump is almost a crime :-S 

For nobody is using this: It is publicly exposed, even via RPC .... If we want more flexibility in changing some internal things, we should not expose them or the very least have documentation stating that the provided API is not stable and if you need a stable interface for XYZ ... please file an issue [here](https://github.com/paritytech/polkadot-sdk/issues).

I also don't think that the proper way is too much fuzz. @tdimitrov I think though that the deprecation process is backwards. What we should be doing, is renaming the new method for keeping backwards compatibility. I don't see it is a problem if runtime API functions end up being named `blah_v3`, the node side can still have nice wrappers once upgraded, having a plain `blah` which is calling into `blah_v3`.

My proposal would be:

1. I don't think we ever used this feature yet, thus we should be able to change the semantics without any warts? (New function is renamed, not the old one)
2. Use it here.

I think that should be straight forward. Am I missing something?

Reply:
@tdimitrov We do want people (e.g. different node implementers) to upgrade to the new version: Can we have a deprecation warning for using API version < X?

Reply:
> I don't think we ever used this feature yet, thus we should be able to change the semantics without any warts? (New function is renamed, not the old one)

It's being used in cumulus for `CollectCollationInfo`

Reply:
> I'd much rather not have duplicated runtime APIs if the inputs and outputs are backwards compatible in scale encoding (and especially since the implementations are the same).

Yeah, the code bloat from this duplication of runtime APIs si quite long term and we do have to keep the primitives duplicated as well.

All the changed runtime APIs are:
- binary compatible, so using old node versions will work just fine
- used at the low levels of stack, not used by team building on polkadot
- Work exactly the same way as before in terms of logic

@eskimor if we do what you suggest, what is the real benefit? What happens if we don’t?

Reply:
I don't see why there would be lot of code bloat ... shouldn't we be able to completely hide this at the node side?

Reply:
I dont understand what hide means here. It will exactly like any other new runtime API you have to check runtime version before deciding if you can call it. Then you’ll have 2 different versions of primitives to deal with in the response.

Reply:
We do have a `From` implementation, you can completely hide that.

Reply:
This unversioned reinterpretation is speced via the RFC and becomes mandatory for node implementers once the corresponding feature bit gets set.